### PR TITLE
[Enhancement] Answer re-commit of a count-evicted transaction from a terminal-state cache

### PR DESCRIPTION
**Pull request by srihithg**
## Why I'm doing:

**Part 1 of 2 addressing a shared-data (lake) Flink backfill failure.**

`label_keep_max_num` count-based transaction eviction **ignores age**, so a finished
transaction can be removed long before `label_keep_max_second`. When a connector
(e.g. Flink) resumes from a savepoint after a long pause and re-commits such a
transaction, the FE returns `transaction not found`, which the connector cannot
distinguish from "never committed" — causing an abort/failover loop.

## What I'm doing:

Keep a bounded per-database cache of terminal (`VISIBLE`/`ABORTED`) transaction outcomes,
populated in `clearTransactionState()` before the full `TransactionState` is dropped. A
re-commit or status probe of an already-evicted transaction is now answered definitively:

- cached `VISIBLE`/`COMMITTED` → idempotent success
- cached `ABORTED` → `TransactionCommitFailedException` (never a false success — critical so a
  truncate-then-resume cannot silently drop data)
- not cached → unchanged `TransactionNotFoundException`

Lookups work by transaction id and by label; `getTxnState`/`getLabelState` also fall back to it.
No edit-log/replay change is needed: eviction runs independently on every FE, so the cache is
populated on every FE (including a future leader).

Also adds config `transaction_terminal_state_cache_num` (default 50000, `0` disables), metrics
`txn_count_eviction` and `txn_terminal_cache_hit`, and a WARN when count-eviction removes a
transaction younger than `label_keep_max_second`.

This is the re-commit-loop half of the backfill failure. The companion PR fixes the
publish-cost half (persistent-index rebuild accounting): #76222. Together the two PRs address
the backfill failure.

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Does this PR entail a change in behavior?
- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

## If yes, please specify the type of change:
- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [x] Parameter changes: default values, similar parameters but with different default values
- [x] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

Transactions removed by count-based eviction now report their real terminal outcome (VISIBLE, COMMITTED or
ABORTED) from `get_load_state` and label-status queries, where they previously reported UNKNOWN. A client
that treated UNKNOWN as "the transaction is gone" now sees the true outcome instead.

Concretely, the Flink connector's lingering-transaction abort now fails closed when a job restores from a
checkpoint older than the one it actually committed. Previously that case proceeded and could silently
duplicate data, so the new behavior is the safe one, but it is visible to anyone who was unknowingly
relying on the old masking. The remediation for an affected job is a new `sink.label-prefix`.

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [x] This PR needs user documentation (for new or modified features or behaviors)
  - [x] I have added documentation for my new feature or new function
- [ ] This is a backport PR

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the PR will be auto-backported to the target branch
  - [ ] 4.1
  - [ ] 4.0
  - [ ] 3.5

### CODE REVIEW

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by kevincai** [comment]
have a design doc of your solution and pros and cons before the implementation. @srihithg

File: fe/fe-core/src/main/java/com/starrocks/transaction/TxnTerminalStateCache.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+    private boolean valid(Record r) {
+        if (r == null) {
+            return false;
+        }
+        if (r.finishTime <= 0) {
+            // Finish time unknown: fall back to the LRU capacity bound only.
+            return true;
+        }
+        return (System.currentTimeMillis() - r.finishTime) / 1000 <= Config.label_keep_max_second;
```
[the remark this answers is not in the document]
Comment (srihithg) [reply fe/fe-core/src/main/java/com/starrocks/transaction/TxnTerminalStateCache.java:246]: Done in 10bb7d7ed0e — `valid()` now returns false when `transaction_terminal_state_cache_num <= 0`, so reads return nothing when the cache is disabled at runtime (even for already-admitted entries). Added `testTerminalCacheDisableAfterPopulatedStopsReads`.

File: docs/en/administration/management/FE_parameters/user_query_loading.md [hunk 2]
```diff
@@ -1070,7 +1070,16 @@ Starting from version 3.3.0, the system defaults to refreshing one partition at
...
+- Introduced in: -
+
+### `transaction_terminal_state_cache_num`
+
+- Default: 50000
+- Type: Int
+- Unit: -
+- Is mutable: Yes
+- Description: The maximum number of terminal (VISIBLE/ABORTED) transaction outcomes kept per database in a lightweight cache after the full transaction state has been evicted by `label_keep_max_num` count-based eviction. Because count-based eviction ignores age, a transaction can be removed long before `label_keep_max_second`. When a connector (for example, Flink) resumes from a savepoint and re-commits such a transaction, the FE would otherwise return "transaction not found", which the connector cannot distinguish from "never committed", causing an abort/failover loop. This cache lets the FE answer the re-commit with the real outcome: idempotent success for VISIBLE/COMMITTED, commit failure for ABORTED. Each entry keeps only the transaction id, label, status, reason, and finish time; entries older than `label_keep_max_second` are treated as absent. Set to `0` to disable the cache.
```
[the remark this answers is not in the document]
Comment (srihithg) [reply docs/en/administration/management/FE_parameters/user_query_loading.md:1154]: Done in 6f3498a627b — added the Japanese translations: `docs/ja/.../FE_parameters/user_query_loading.md` (new config + count-eviction-ignores-age note) and `docs/ja/.../monitoring/metric_details/t-z.md` (`txn_count_eviction` / `txn_terminal_cache_hit`).

File: fe/fe-core/src/main/java/com/starrocks/transaction/DatabaseTransactionMgr.java [hunk 3]
```diff
@@ -476,7 +489,13 @@ public VisibleStateWaiter commitPreparedTransaction(long transactionId) throws S
 
         TransactionState transactionState = getTransactionState(transactionId);
         if (transactionState == null) {
-            throw new TransactionNotFoundException(transactionId);
+            // The full state may have been evicted (count-based eviction ignores age). If the
+            // terminal outcome is cached and was VISIBLE, treat the re-commit as idempotent success
+            // and hand back an already-completed waiter; if it was ABORTED, fail the commit.
+            resolveEvictedTxnOrThrow(transactionId);
+            LOG.info("transaction {} was already evicted with terminal state VISIBLE, "
```
[the remark this answers is not in the document]
Comment (srihithg) [reply fe/fe-core/src/main/java/com/starrocks/transaction/DatabaseTransactionMgr.java:497]: Fixed in 1ad58936494. You were right — the 2PC entry points (`GlobalTransactionMgr.prepareTransaction`/`commitPreparedTransaction`, and `commitTransactionUnderDatabaseWLock`) deref `getTransactionState(...).getTableIdList()` before delegating, so an evicted txn NPE'd before reaching the cache fallback. All three sites now null-guard and delegate straight to the db mgr (no tables to lock for a cache-only resolution). Added `testRecommitEvictedTxnViaGlobalTransactionMgr2PCPath` which reproduces the NPE on the real GlobalTransactionMgr path and passes with the fix.

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

File: fe/fe-core/src/main/java/com/starrocks/transaction/DatabaseTransactionMgr.java [hunk 4]
```diff
@@ -2035,12 +2104,25 @@ public void removeExpiredTxns(long currentMillis) {
...
+            }
         } finally {
             writeUnlock();
         }
     }
 
     private void clearTransactionState(TransactionState transactionState) {
+        // Remember the terminal outcome before the full state is dropped, so a later re-commit of
+        // this (now evicted) transaction can be answered definitively instead of "not found".
+        terminalStateCache.put(transactionState);
```
[the remark this answers is not in the document]
Comment (srihithg) [reply fe/fe-core/src/main/java/com/starrocks/transaction/DatabaseTransactionMgr.java:2165]: This is a deliberate design choice rather than an oversight. The cached outcome is always derived from the same replicated TransactionState that every FE already agrees on, and clearTransactionState populates the cache on each FE during normal replay (every FE runs its own labelCleaner), so a follower promoted to leader still has it. The only uncovered case is an FE that restarts after eviction and rebuilds from an image that no longer contains the transaction. We accept that because it degrades safely: the worst case is a freshly restarted leader answering not found where a long running one would have answered committed, and the connector then retries and succeeds. It never returns a wrong answer.

Persisting the outcome would add a new FE image and editlog format for a window that already fails safe, which we deliberately avoided to keep this bugfix backport safe. If the restart window turns out to matter in practice we can add durable persistence as a separate main only change.

**Comment by CelerData-Reviewer** [comment]
@codex review

File: fe/fe-core/src/main/java/com/starrocks/transaction/DatabaseTransactionMgr.java [hunk 5]
```diff
@@ -158,6 +158,14 @@ public class DatabaseTransactionMgr {
     private final Map<String, Set<Long>> labelToTxnIds = Maps.newHashMap();
     private long maxCommitTs = 0;
 
+    /*
+     * Lightweight cache of terminal (VISIBLE/ABORTED) transaction outcomes that survives the
+     * removal of the full TransactionState by count-based eviction (removeExpiredTxns). Lets a
+     * connector re-commit of an already-cleaned transaction get the real outcome instead of a
+     * "transaction not found" error. See TxnTerminalStateCache.
+     */
+    private final TxnTerminalStateCache terminalStateCache = new TxnTerminalStateCache();
```
[the remark this answers is not in the document]
Comment (srihithg) [reply fe/fe-core/src/main/java/com/starrocks/transaction/DatabaseTransactionMgr.java:167]: Good catch, and this one changes the calculus versus the earlier best-effort note. I verified GlobalStateCheckpointWorker runs clearExpiredJobs() (which count-evicts and populates the in-memory cache while dropping the full TransactionState) before saveImage(), so the checkpoint image systematically lacks both the transaction and its cached outcome, and any FE that loads it returns not found. That is broader than the restart-only window I described before, so I implemented persistence.

The terminal-state cache is now serialized into the GLOBAL_TRANSACTION_MGR image block as a trailing, count-prefixed collection and repopulated on load. It is forward and backward compatible: older images lack the section (the EOF is caught and skipped on read), and an older FE reading a newer image discards the trailing records in reader.close(). restore() applies the same admission rules as put(), so a stale entry in an old image is dropped. Added a unit round-trip test plus an end-to-end image save/load test asserting a follower resolves an evicted outcome from the restored cache. Pushed in 9dc0c31.

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

File: fe/fe-core/src/main/java/com/starrocks/http/rest/TransactionLoadAction.java [hunk 6]
```diff
@@ -313,7 +315,17 @@ private TransactionOperationHandler getTxnOperationHandler(TransactionOperationP
...
+                // terminal outcome is still known (terminal-state cache), route to the default handler,
+                // which returns the idempotent terminal result for a savepoint/resume recommit;
+                // otherwise the label is genuinely unknown.
+                TransactionStateSnapshot snapshot = GlobalStateMgr.getCurrentState()
+                        .getGlobalTransactionMgr().getLabelStatus(db.getId(), label);
+                if (snapshot.getStatus() == TransactionStatus.UNKNOWN) {
+                    throw new StarRocksException(
+                            String.format("No transaction found with db[%s] and label[%s]", dbName, label));
+                }
+                return new TransactionWithoutChannelHandler(params);
```
[the remark this answers is not in the document]
Comment (srihithg) [reply fe/fe-core/src/main/java/com/starrocks/http/rest/TransactionLoadAction.java:328]: Good catch, fixed. TXN_PREPARE in this handler previously no-op'd and fell through to a BE redirect, so an evicted transaction retried with PREPARE (routed here even for a bypass-write client, since source_type is omitted on retry and the cache cannot recover it) would redirect through the now-expired coordinator cache and fail. It now answers the retry idempotently from the terminal-state cache when getLabelStatus reports a terminal outcome (VISIBLE/COMMITTED/ABORTED); PREPARED (live) and UNKNOWN still fall through to the redirect, so the ordinary stream-load prepare path is unchanged. It reads only getLabelStatus, so it does not disturb the redirect path's own getLabelTransactionState call. Pushed in 5534f26.

File: fe/fe-core/src/main/java/com/starrocks/transaction/TxnTerminalStateCache.java [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+        if (existing == null || txnId >= existing.txnId) {
+            byLabel.put(label, r);
+        }
+    }
+
+    /**
+     * @return a snapshot of the currently cached records, for FE image serialization.
+     */
+    synchronized List<Record> snapshot() {
+        return new ArrayList<>(byTxnId.values());
```
[the remark this answers is not in the document]
Comment (srihithg) [reply fe/fe-core/src/main/java/com/starrocks/transaction/TxnTerminalStateCache.java]: Good catch. Because byTxnId and byLabel are independent access-order LRUs, a label-hot record can outlive its byTxnId entry, and snapshot() only serialized byTxnId.values(), so a checkpoint could silently drop it and regress getLabelState to UNKNOWN. snapshot() now returns the union of both indexes, deduped by txn id. Added a test that forces a byTxnId eviction while the label stays hot and asserts the record is still present in the snapshot. Pushed in 5534f26.

**Comment by CelerData-Reviewer** [comment]
@codex review

File: fe/fe-core/src/main/java/com/starrocks/transaction/TxnTerminalStateCache.java [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+
+    TxnTerminalStateCache() {
+        byTxnId = new LinkedHashMap<Long, Record>(16, 0.75f, true) {
+            @Override
+            protected boolean removeEldestEntry(Map.Entry<Long, Record> eldest) {
+                int capacity = Config.transaction_terminal_state_cache_num;
+                return capacity <= 0 || size() > capacity;
+            }
+        };
+        byLabel = new LinkedHashMap<String, Record>(16, 0.75f, true) {
```
[the remark this answers is not in the document]
Comment (srihithg) [reply fe/fe-core/src/main/java/com/starrocks/transaction/TxnTerminalStateCache.java]: Good catch, fixed. Collapsed the two independent LRUs into a single canonical one: byTxnId is now the only Record store (capped at transaction_terminal_state_cache_num), and label lookups go through a lightweight label -> latest txn id pointer index rather than a second Record map. getByLabel resolves via byTxnId.get, which touches the record in the access-order LRU, so a label-hot outcome stays resident under the single cap instead of aging out as if unused; the pointer is pruned when its target is evicted, so it never dangles. snapshot() is now exactly byTxnId.values(), so total retention and the checkpoint are bounded to the configured maximum, not ~2x. Added a test asserting retention never exceeds the cap, and one that a label-hot record survives eviction while an un-accessed one is dropped. Pushed in a49323f.

File: fe/fe-core/src/main/java/com/starrocks/transaction/TxnTerminalStateCache.java [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+            return;
+        }
+        Record r = new Record(txnId, label, status, reason, finishTime);
+        byTxnId.put(txnId, r);
+        // Largest txn id wins for a reused label, matching getLabelState()'s "largest id" semantics.
+        Long cur = labelToTxnId.get(label);
+        if (cur == null || txnId >= cur) {
+            labelToTxnId.put(label, txnId);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (srihithg) [reply fe/fe-core/src/main/java/com/starrocks/transaction/TxnTerminalStateCache.java]: Fixed in f95be11a329. \`removeEldestEntry\` only trims one entry per insert, so a runtime reduction of \`transaction_terminal_state_cache_num\` left the cache above the new bound. \`insert()\` (and \`snapshot()\`) now call \`trimToCapacity()\`, which clears when disabled and evicts eldest entries until \`size() <= cap\`. Covered by \`testRuntimeCapReductionDrainsToNewCap\`.

File: fe/fe-core/src/main/java/com/starrocks/transaction/TxnTerminalStateCache.java [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+     * resident under the single capacity rather than aging out as if unused.
+     */
+    synchronized Record getByLabel(String label) {
+        Long txnId = labelToTxnId.get(label);
+        if (txnId == null) {
+            return null;
+        }
+        Record r = byTxnId.get(txnId);
+        return valid(r) ? r : null;
+    }
```
[the remark this answers is not in the document]
Comment (srihithg) [reply fe/fe-core/src/main/java/com/starrocks/transaction/TxnTerminalStateCache.java:231]: Fixed in f95be11a329. \`getByTxnId\`/\`getByLabel\` resolve through \`byTxnId.get\`, which touches the record to MRU before \`valid()\` can reject it; an expired (or disabled-cache) record is now removed on access via \`removeRecord\` instead of lingering and evicting valid entries. Covered by \`testExpiredRecordPrunedOnAccess\`.

File: fe/fe-core/src/main/java/com/starrocks/common/Config.java [hunk 11]
```diff
@@ -512,6 +512,23 @@ public class Config extends ConfigBase {
...
+     * long before *label_keep_max_second*. When a connector (e.g. Flink) resumes from a savepoint
+     * and re-commits such a transaction, the FE would otherwise return "transaction not found",
+     * which the connector cannot distinguish from "never committed", causing an abort/failover loop.
+     * This cache lets the FE answer the re-commit with the real outcome (idempotent success for
+     * VISIBLE/COMMITTED, commit-failed for ABORTED). Each entry is tiny (id, label, status, reason,
+     * finish time); entries older than *label_keep_max_second* are treated as absent. Set to 0 to
+     * disable the cache.
+     */
+    @ConfField(mutable = true)
+    public static int transaction_terminal_state_cache_num = 50000;
```
[the remark this answers is not in the document]
Comment (srihithg) [reply fe/fe-core/src/main/java/com/starrocks/common/Config.java:614]: Updated the PR metadata: this is an intentional behavior change (a re-commit of a count-evicted transaction now returns its real terminal outcome instead of "transaction not found"), so the behavior-change classification is now set to Yes with Policy changes checked.

File: fe/fe-core/src/main/java/com/starrocks/http/rest/transaction/BypassWriteTransactionHandler.java [hunk 12]
```diff
@@ -221,11 +234,16 @@ private TransactionResult handleRollbackTransaction(Database db,
...
+    // (e.g. count-evicted after a savepoint/resume). A null return signals the caller to resolve the
+    // terminal outcome from the terminal-state cache (commit*/rollbackEvictedByLabel) instead of
+    // failing with "transaction not found". Throws only when a live transaction exists but was not
+    // created by a bypass write.
     private static TransactionState getTxnState(long dbId, String label) throws StarRocksException {
         TransactionState txnState = GlobalStateMgr.getCurrentState()
                 .getGlobalTransactionMgr().getLabelTransactionState(dbId, label);
         if (null == txnState) {
-            throw new StarRocksException(String.format("No transaction found by label %s", label));
+            return null;
```
Comment (luohaha) [inline fe/fe-core/src/main/java/com/starrocks/http/rest/transaction/BypassWriteTransactionHandler.java:246]: **[P2] Preserve source-type validation for cached outcomes**

While the full `TransactionState` is present, this handler rejects a label whose transaction was not created by `BYPASS_WRITE` (lines 249–254). After count eviction, however, this null return routes the request to `commitEvictedByLabel`/`rollbackEvictedByLabel`, which accepts any cached terminal outcome for the label because the cache/image record does not retain `LoadJobSourceType`. Consequently, an explicit `source_type=BYPASS_WRITE` prepare/commit can return success for an evicted transaction from another source even though the identical request is rejected before eviction. Please retain the source type in the cached/image record and validate it here, or otherwise avoid using an untyped cached outcome for an explicitly typed bypass-write request.

File: fe/fe-core/src/main/java/com/starrocks/transaction/TxnTerminalStateCache.java [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+
+    /**
+     * @return a snapshot of the currently cached records, for FE image serialization. This is exactly
+     * byTxnId (the single Record store), so the snapshot is complete and bounded by the configured
+     * capacity -- there is no second Record index to union in, and label lookups resolve through
+     * byTxnId, so nothing is retained outside it.
+     */
+    synchronized List<Record> snapshot() {
+        trimToCapacity(); // never serialize more than the configured cap into the image
+        return new ArrayList<>(byTxnId.values());
```
Comment (luohaha) [inline fe/fe-core/src/main/java/com/starrocks/transaction/TxnTerminalStateCache.java]: **[P2] Prune expired records before snapshotting**

`snapshot()` only enforces the capacity bound; the age check in `valid()` runs only when a record is looked up. A record that has aged past `label_keep_max_second` but is never queried therefore remains in memory and is serialized into every FE image until capacity pressure evicts it or the FE restarts (where `restore()` finally drops it). With the default cap of 50,000 per database, inactive databases can contribute a large amount of permanently unusable checkpoint data. Please remove age-expired records before copying `byTxnId.values()` (or prune them from the periodic cleaner), and add a snapshot-after-expiry regression test.

**Review by luohaha — COMMENTED** [review]
Two remaining edge cases in the terminal-state cache path:

**Review by luohaha — DISMISSED** [review]

**Comment by CelerData-Reviewer** [comment]
@codex review

File: fe/fe-core/src/main/java/com/starrocks/transaction/GlobalTransactionMgr.java [hunk 14]
```diff
@@ -371,9 +382,21 @@ public void commitPreparedTransaction(Database db, long transactionId, long time
...
+            // tables to lock; resolve the outcome from the terminal-state cache. commitPreparedTransaction
+            // returns an already-visible waiter for a cached VISIBLE outcome and throws for
+            // ABORTED/not-found. This is the Flink savepoint/resume re-commit path.
+            getDatabaseTransactionMgr(db.getId()).commitPreparedTransaction(transactionId);
+            MetricRepo.COUNTER_LOAD_FINISHED.increase(1L);
+            return;
+        }
+        // Live transaction: use upstream's rate-limit-aware retry path (lock + commit under the intensive
+        // db lock, retrying on CommitRateExceededException within the timeout budget).
         waiter = retryCommitPreparedOnRateLimitExceeded(db, transactionId, timeoutMillis);
```
[the remark this answers is not in the document]
Comment (srihithg) [reply fe/fe-core/src/main/java/com/starrocks/transaction/GlobalTransactionMgr.java]: Fixed in b62b50332dd. Confirmed the race: `commitPreparedTransactionUnderIntensiveDbLock` re-reads via `getTransactionStateOrThrow` before locking, so a count-eviction between the pre-check and that lookup made the recommit throw `TransactionNotFoundException` even though the terminal cache held the outcome. The live path now catches `TransactionNotFoundException` and falls back to `DatabaseTransactionMgr.commitPreparedTransaction`, which resolves from the cache (idempotent success for VISIBLE/COMMITTED, commit-failed for ABORTED) and re-throws only for a genuinely absent transaction. Added a spy-based regression test (`testCommitPreparedFallsBackToCacheWhenEvictedMidCommit`).

**Comment by CelerData-Reviewer** [comment]
@codex review

File: fe/fe-core/src/main/java/com/starrocks/http/rest/transaction/TransactionWithoutChannelHandler.java [hunk 15]
```diff
@@ -147,13 +186,47 @@ private TransactionResult handleRollbackTransaction(Database db,
...
+    // genuinely unknown -> not found. Mirrors the in-map status branches in handleCommitTransaction.
+    private static TransactionResult commitEvictedTransactionByLabel(long dbId, String label) throws StarRocksException {
+        TransactionStateSnapshot snapshot =
+                GlobalStateMgr.getCurrentState().getGlobalTransactionMgr().getLabelStatus(dbId, label);
+        TransactionResult result = new TransactionResult();
+        switch (snapshot.getStatus()) {
+            case COMMITTED:
+            case VISIBLE:
+                result.setOKMsg(String.format("Transaction has already committed, label is %s", label));
+                result.addResultEntry(TransactionResult.LABEL_KEY, label);
```
[the remark this answers is not in the document]
Comment (srihithg) [reply fe/fe-core/src/main/java/com/starrocks/http/rest/transaction/TransactionWithoutChannelHandler.java:201]: Fixed in e17760e1170. Valid finding — and the identical gap existed symmetrically in `BypassWriteTransactionHandler` (its evicted commit/rollback success branches), so I fixed all four eviction-recovery success branches across both handlers.

The transaction id is now carried through `TransactionStateSnapshot`: the terminal-state cache `Record` already stores it (it's the cache's key) and the image persists it via `@SerializedName("txnId")`, so `getLabelState`/`getTxnState` thread it into the snapshot. Both handlers echo `TxnId` alongside `Label` via a shared `TransactionOperationHandler.addCachedTxnId` helper, so the eviction-recovery response now matches the live path and the documented response shape at `docs/en/loading/Stream_Load_transaction_interface.md:411` and `:432`. The helper guards with `txnId > 0`, which also suppresses a GSON-defaulted `0`. Added e2e assertions covering all four branches (commit + rollback, bypass + without-channel).

**Review by luohaha — DISMISSED** [review]

**Comment by kevincai** [comment]
Giving the size of the code change, it is beyond a bugfix. Any alternatives we can think of if willing to carry the fix to v4.0 and v3.5? @srihithg

**Review by xiangguangyxg — DISMISSED** [review]

**Review by kevincai — COMMENTED** [review]
I would suggest a design doc and design review before merging this commit.

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review
